lib/Solve.js

```diff
@@ -0,0 +1,72 @@
+// @flow
+
+const spawn = require('cross-spawn');
+
+const extraPackages = ['elm/core', 'elm/json', 'elm/time', 'elm/random'];
+
+function getDependencies(
+  pathToElmJson /*: string */,
+  projectElmJson /*: any */
+) /*: { direct: { [string]: string }, indirect: { [string]: string } } */ {
+  // If possible, use a simple and fast merge of all dependencies in elm.json,
+  // otherwise use elm-json. At the time of writing, this could save 1 second in
+  // an application project.
+  const simpleMerge = trySimpleMerge(projectElmJson);
+  if (simpleMerge !== undefined) {
+    return simpleMerge;
+  }
```

Comment:
@zwilias As our elm dependencies expert, would you mind checking if `trySimpleMerge` is sound, or if it can cause errors? If it works this is a pretty great speedup.

Reply:
You should be able to merge them if they are unique indeed, but under the assumption that indirect dependencies and indirect test dependencies are correct. This is fairly often the case for project indirect dependencies since they are checked by elm. But for the indirect test dependencies, it is a bit more difficult to trust (especially for people not using `elm-json` to install their test dependencies).

Reply:
I don’t think it matters much if dependencies are invalid because the error you get is OK:

```
~/forks/node-test-runner/example-application js *$
❯ ../bin/elm-test
-- INCOMPATIBLE DEPENDENCIES ------------------------------------------ elm.json

The dependencies in your elm.json are not compatible.

Did you change them by hand? Try to change it back! It is much more reliable to
add dependencies with elm install or the dependency management tool in
elm reactor.

Please ask for help on the community forums if you try those paths and are still
having problems!
```

It’s basically the same error as if you try to compile your project with `elm make`.

Reply:
This is the error you get (from elm-json) if disabling my shortcut:

```
~/forks/node-test-runner/example-application js *$
❯ ../bin/elm-test

-- NO VALID PACKAGE VERSION ----------------------------------------------------

Because elm-explorations/test 1.2.0 depends on elm/virtual-dom 1.0.0 <= v <
2.0.0 and elm/virtual-dom 1.0.0 <= v < 2.0.0 is unavailable, no valid set of
package versions could be found.

And because this project depends on elm-explorations/test 1.2.0, no valid set of
package versions could be found.
```

Not sure it’s much better?

Reply:
This is because elm-json is used in offline mode right? does it do the same in online mode?

Reply:
As far as I know, elm-json is only in offline mode if you use the `--offline` flag. That’s not present in this file. And I’m connected to the Internet. But maybe my example is bad. I made the elm.json invalid by saying I depend on elm/virtual-dom 2.0.0 which doesn’t exist.

Reply:
I see, maybe try removing an indirect dependency to see if there is a difference

Reply:
I removed elm/virtual-dom from indirect. Elm fails with the usual message, while `elm-json solve` adds it back, making the tests run. But if you’d try to compile your code you’ll of course get Elm’s error. So maybe it’s weird that the tests work, when compiling the project doesn’t? Or rather – I don’t think it matters that the fast version doesn’t make it since your project is kind of invalid anyway.

Reply:
Oh yeah sorry I meant remove something from indirect test dependencies. Since most often (before elm-json) people would just add some library to their test dependencies if they need them for tests, but would not search for the indirect ones. In that case, elm-test used to work, and works with elm-json. But it could make sense to fail if the dependencies are incorrect.

Reply:
Ok, I used `elm-test install truqu/md5` which correctly added that to direct test-dependencies, as well as `zwilias/elm-utf-tools` to _indirect_ test-dependencies. I added a `MD5.hex` call to the tests just to make sure everything worked.

Then i removed `zwilias/elm-utf-tools` from elm.json. Same thing as before – `elm-json solve` adds it back, so tests pass, while `elm make` or my fast version fail with the “were edited by hand” error.